### PR TITLE
Feature/logistic regression

### PR DESCRIPTION
**Pull request by lindechen**
The **base_model** folder contains two files: `lasso.R` and `logit.R`. There are two outcome variables being examined: `grad` - a continuous variable (see codebook) and `educ_cat` - a categorical variable with 4 levels. 

- `lasso.R` - performs LASSO penalty regression on demographic variables chosen by @CarolinaVelasco for variable selection.

- `logit.R` - performs **linear regression** on the 1) chosen demographic variables from `lasso.R` and 2) housing quality index created by @CarolinaVelasco on `grad`; performs **multinomial logistic regression** on the 1) chosen demographic variables and 2) housing quality on `educ_cat`.

The 6 components used to derive the index are not used. 

Please review this pull request. Thank you.

### CODE REVIEW

File: base_model/lasso.R [hunk 1]
```diff
@@ -0,0 +1,55 @@
+library(glmnet) # load 'glmnet' package
+
+data <- readRDS("./Data/df_index_demographic_vars.RDS") # read index and demographics data provided by Carolina
+
+data <- data[,-1] # remove 'control', an ID variable
+data$sex <- data$sex-1 # make 'sex' a 0/1 binary variable
+data$data_year <- data$data_year-1997 # normalize 'data_year' to start at 1997
```
Comment (CarolinaVelasco) [inline base_model/lasso.R]: The data years are 1997-2013, every two years. Does this mean that the possible values for year are: 0 (1997), 2 (1999), 4 (2001), and so on?

File: lasso.R [hunk 2]
```diff
@@ -0,0 +1,55 @@
+library(glmnet) # load 'glmnet' package
+
+data <- readRDS("./Data/df_index_demographic_vars.RDS") # read index and demographics data provided by Carolina
```
Comment (jcochanc) [inline lasso.R]: Where is this data? Can you show me a screenshot? For some reason, my branch does not have this data. If this is the data from @CarolinaVelasco's PR, I don't know how it is being used here since it hasn't been merged into Master. Sorry im so confused!

File: lasso.R [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+data <- data[,-1] # remove 'control', an ID variable
+data$sex <- data$sex-1 # make 'sex' a 0/1 binary variable
+data$data_year <- data$data_year-1997 # normalize 'data_year' to start at 1997
+data$weight <- (data$weight - mean(data$weight)) / sd(data$weight) # standardize `weight`
+
+######################### perform LASSO regression on demographic variables #####################
+# Note: LASSO inherently performs variable selection by shrinking some coefficients to zero
+
+# 1. try with 'grad' (see codebook) - CONTINOUS OUTCOME VARIABLE
+x <- as.matrix(data[,-c(5,6,24,25,26,27,28,29,30)]) # Removes class, educ_cat (redundant), dimension vars, index (both non-demographic)
```
Comment (jcochanc) [inline lasso.R]: @lindechen Try to avoid using index values. As we found out, the data isn't finalized and if one little thing changes when @CarolinaVelasco 's PR comes in, all these indices you are pointing to will be shifted. Dplyr has some great piping that lets your code be legible and it won't be subject to index shifts. For example:

x <- data %>% select(variable1, variable2,..., variableN)

File: lasso.R [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+y <- as.double(as.matrix(data[, 5])) # Only class
```
Comment (jcochanc) [inline lasso.R]: See comment above.

File: lasso.R [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+# Fitting the model (Lasso: alpha = 1)
+set.seed(14)
+cv.lasso <- cv.glmnet(x, y, alpha=1, parallel=TRUE, standardize=TRUE, type.measure='mse')
```
Comment (jcochanc) [inline lasso.R]: I thought parallel had to be FALSE for reproducible results?

File: lasso.R [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+# all coef
+coef <- coef(cv.lasso, s=cv.lasso$lambda.1se)[,1] # add 1 SE to obtain more parsimonious model
+# non-zero demo. coef
+nonzero_coef_grad <- data.frame(coef = coef[which(coef(cv.lasso, s=cv.lasso$lambda.1se) != 0)])  
+
+saveRDS(nonzero_coef_grad, "./Data/nonzero_coef_grad.RDS")
+
+# 2. try with `educ_cat` - MULTINOMIAL OUTCOME VARIABLE
+x <- as.matrix(data[,-c(5,6,24,25,26,27,28,29,30)]) # Removes class, grad (redundant), dimension vars, index (both non-demographic)
```
Comment (jcochanc) [inline lasso.R]: Same comment; this is subject to index variation and should be able to run on the same data with columns shifted around.

File: lasso.R [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+y <- as.matrix(as.numeric(data[, 6]))
```
Comment (jcochanc) [inline lasso.R]: Same as above.

File: lasso.R [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+index <- c() # create empty vector to store var. with coefficients in any of 4 categories shrunk to zero
+for (i in 1:nrow(coef)) {
+  if (sum(coef[i,] == 0) > 0) {
+    index = c(index, i)
+  }
+}
+
+# non-zero demo. coef
+nonzero_coef_educ_cat <- coef[-index,] 
+saveRDS(nonzero_coef_educ_cat, "./Data/nonzero_coef_educ_cat.RDS")
```
Comment (jcochanc) [inline lasso.R]: So this wasn't run or pushed? I don't see it in data.

File: logit.R [hunk 9]
```diff
@@ -0,0 +1,30 @@
+library(nnet)
+
+data <- readRDS("./Data/df_index_demographic_vars.RDS") # provided by Carolina in pca for index
+vars_grad <- readRDS("./Data/nonzero_coef_grad.RDS") # demo. var. by LASSO using `grad` as outcome
+vars_educ_cat <- readRDS("./Data/nonzero_coef_educ_cat.RDS") # demo. var. by LASSO using `educ_cat` as outcome
+df <- data[,-c(24,25,26,27,28,29)] # remove 6 components to create df for regression // use index only
```
Comment (jcochanc) [inline logit.R]: Use component names and pipe it so we know which are being pulled.

File: logit.R [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+# Perform linear regression on vars_grad (continuous outcome)
+lm_vars <- rownames(vars_grad)[-1]
+df_grad <- df[,c("grad", lm_vars, "housing_qual_index")]
+linear_mod <- lm(grad ~., df_grad)
+summary_lm <- summary(linear_mod)$coefficients # summary table for linear regression
+
+saveRDS(summary_lm, "./Data/summary_lm.RDS")
```
Comment (jcochanc) [inline logit.R]: I don't see this in the data folder.

File: logit.R [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+# calculate p-values
+multinom_coef <- summary(multinom_mod)$coefficients
+z <- multinom_coef/summary(multinom_mod)$standard.errors
+p <- round((1 - pnorm(abs(z), 0, 1))*2,4) # p-value table for ordinal logistic regression
+
+summary_multinom <- data.frame(t(multinom_coef), t(p)) # summary table for ordinal logistic regression
+colnames(summary_multinom) <- c("level 1 coef.", "level 2 coef.", "level 3 coef.", 
+                                "level 1 p-val.", "level 2 p-val.", "level 3 p-val.")
+
+saveRDS(summary_multinom, "./Data/summary_multinom.RDS")
```
Comment (jcochanc) [inline logit.R]: Data not found.

**Review by jcochanc — CHANGES_REQUESTED** [review]

**Comment by jcochanc** [comment]
@lindechen I just realized I didn't see you partition the data into train and test; was this done elsewhere? I know you have a CV step in there but I wonder how we are going to compare the models (logit VS xgboost). If you want, and it wasnt done, I can partition that data once the PR is in from #12

File: base_model/lasso.R (hunk 1 above)
Comment (lindechen) [reply base_model/lasso.R]: That is correct! 1997 would be year 0, 1999 would be year 2, etc.

File: lasso.R (hunk 2 above)
Comment (lindechen) [reply lasso.R]: This will be resolved when the data is in master

File: lasso.R (hunk 3 above)
Comment (lindechen) [reply lasso.R]: That is a good point. I will make the change!

File: lasso.R (hunk 5 above)
Comment (lindechen) [reply lasso.R]: That's possible; I will look into this.

**Comment by lindechen** [comment]
@jcochanc No, there wasn't a train/test set and if you could create one, that would be great! 

For my part, I can then perform CV on the training set and regress on the test set. That way we can compare the base model(s) with XGBoost.

**Comment by jcochanc** [comment]
Okay I’ll make an issue for it and the. Have the done as soon as the PR is approved for Carolina.

File: base_model/lasso.R [hunk 12]
```diff
@@ -0,0 +1,55 @@
+library(glmnet) # load 'glmnet' package
+
+data <- readRDS("./Data/df_index_demographic_vars.RDS") # read index and demographics data provided by Carolina
+
+data <- data[,-1] # remove 'control', an ID variable
+data$sex <- data$sex-1 # make 'sex' a 0/1 binary variable
```
Comment (jcochanc) [inline base_model/lasso.R]: Chris Schmidt got me in the habit of changing these kind of variables to say male or female so that the 0-1 has some meaning. That way when you see the variable female==1 you know what the means. 

Also try using the ifelse() function. This case may work because you checked the format type and it was int or numeric but ifelse() would avoid that issue since you’d be forced to check the type while doing the change. Just a suggestion! I love ifelse().

**Comment by lindechen** [comment]
> Okay I’ll make an issue for it and the. Have the done as soon as the PR is approved for Carolina.

Great!

File: lasso.R (hunk 6 above)
Comment (lindechen) [reply lasso.R]: this is resolved in the updated file.

**Comment by lindechen** [comment]
I submitted updated `lasso.R` and `logit.R` files that:

- resolve the problems above
- use the train/test data built from the alternative index
- perform LASSO regression for variable selection
- use multinomial (ordinal) logistic regression to predict the `educ_cat` on the test set *with* and *without* the alternative index.

Please review this PR.

**Review by jcochanc — APPROVED** [review]
Great job, Linde!
